GitCommands/Repository/RepositoryManager.cs

```diff
@@ -123,5 +104,40 @@ public static Task SaveRepositoryRemoteHistoryAsync(RepositoryHistory repository
                 RepositoryStorage.Save(KeyRemoteHistory, repositoryHistory.Repositories);
             });
         }
+
+        private static Task<RepositoryHistory> AddAsMostRecentRepositoryAsync(string repositoryPath, Func<Task<RepositoryHistory>> loadRepositoryHistory, Func<RepositoryHistory, Task> saveRepositoryHistory)
+        {
+            return Task.Run(async () =>
+            {
+                var repositoryHistory = await LoadRepositoryHistoryAsync();
+
+                if (string.IsNullOrWhiteSpace(repositoryPath))
+                {
+                    return repositoryHistory;
+                }
+
+                AddMostRecentRepository(repositoryHistory.Repositories, repositoryPath);
```

Comment:
@sharwell could you help me out? I'm drawing a mental blank...
![image](https://user-images.githubusercontent.com/4403806/38424851-acb6dcb0-39f5-11e8-9f7d-15ff0909168b.png)

What's the best way of executing the following?
* load repositories asyncronously
* if the supplied path is valid
  - update the list of repositories
  - save the list asyncronously

Reply:
Your code has both `AddMostRecentRepository` and `AddMostRecentRepositoryAsync`. In this case, the behavior is assumed to be the same and the async form must be the one called from an async method. The solution is one of the following two items:

1. Make sure they behave the same and call the appropriate one
2. Rename one of them

Reply:
The analyzer doesn't distinguish between local functions and methods, so local function `AddMostRecentRepository` is treated as the synchronous (and blocking) functional equivalent of method `AddMostRecentRepositoryAsync`.

Reply:
I suppose I can suppress the warning:

```
#pragma warning disable VSTHRD103 // Call async methods when in an async method
                AddMostRecentRepository(repositoryHistory.Repositories, repositoryPath);
#pragma warning restore VSTHRD103 // Call async methods when in an async method
```

...or move the code from the local function into the body of the `Task.Run`:

```
        private static Task<RepositoryHistory> AddAsMostRecentRepositoryAsync(string repositoryPath, Func<Task<RepositoryHistory>> loadRepositoryHistory, Func<RepositoryHistory, Task> saveRepositoryHistory)
        {
            return Task.Run(async () =>
            {
                var repositoryHistory = await LoadRepositoryHistoryAsync();

                if (string.IsNullOrWhiteSpace(repositoryPath))
                {
                    return repositoryHistory;
                }

                repositoryPath = repositoryPath.Trim();

                var repository = repositoryHistory.Repositories.FirstOrDefault(r => r.Path.Equals(repositoryPath, StringComparison.CurrentCultureIgnoreCase));
                if (repository != null)
                {
                    repositoryHistory.Repositories.Remove(repository);
                }
                else
                {
                    repository = new Repository(repositoryPath);
                }

                repositoryHistory.Repositories.Insert(0, repository);

                await SaveRepositoryHistoryAsync(repositoryHistory);

                return repositoryHistory;
            });
        }
```

Reply:
Either way works. I couldn't think of a different name for the local function which is "definitely better", but if you come up with one that would work too.